File: src/main/java/discretemath/structure/sequence/concrete/GeometricProgression.java
package discretemath.structure.sequence.concrete;

import java.math.BigDecimal;

import discretemath.structure.sequence.SimpleSequence;
import discretemath.structure.set.Sets;

public final class GeometricProgression extends SimpleSequence<BigDecimal> {

	private final BigDecimal initialTerm;

	private final BigDecimal ratio;

	public GeometricProgression(BigDecimal initialTerm, BigDecimal ratio) {
		super(Sets.R);
		this.initialTerm = initialTerm;
		this.ratio = ratio;
	}

	@Override
	protected BigDecimal getTerm(int n) {
		return initialTerm.multiply(ratio.pow(n));
	}
}

File: src/main/scala/discretemath/structure/sequence/concrete/GeometricProgression.scala
package discretemath.structure.sequence.concrete

import discretemath.structure.sequence.SimpleSequence
import discretemath.structure.set.Sets

final class GeometricProgression(val initialTerm: BigDecimal, val ratio: BigDecimal) extends SimpleSequence[BigDecimal](Sets.R) {

  override protected def applyWithoutCheckInput(n: BigInt): BigDecimal = initialTerm * ratio.pow(n.intValue)

  override def isOneToOne = true
}